transaction/builder/src/memo_builder/rth_memo_builder.rs

```diff
@@ -76,12 +75,107 @@ pub struct RTHMemoBuilder {
     fee: Amount,
 }
 
+#[derive(Clone, Debug)]
+pub enum CustomMemoType {
+    PaymentRequestId(u64),
+    PaymentIntentId(u64),
+    FlexibleMemoGenerator(FlexibleMemoGenerator),
+}
+
+/// This is a trait that must be implemented by generators used for creating
+/// flexible output memos and flexible change memos
+pub trait FlexibleMemoGeneratorTrait: Sync + Send + Debug {
+    /// This is called when generating output memos.
+    /// It must return a memo payload with the first byte being 0x01
+    fn create_output_memo(
+        &self,
+        context: FlexibleMemoOutputContext,
+    ) -> Result<FlexibleMemoPayload, NewMemoError>;
+    /// This is called when generating change memos.
+    /// It must return a memo payload with the first byte being 0x02
+    fn create_change_memo(
+        &self,
+        context: FlexibleMemoChangeContext,
+    ) -> Result<FlexibleMemoPayload, NewMemoError>;
+}
+
+/// This is a struct that contains a generator used for creating flexible output
+/// memos and flexible change memos
+#[derive(Clone)]
+pub struct FlexibleMemoGenerator {
+    /// This is a reference to the actual generator that will be called.
+    pub generator: Arc<Box<dyn FlexibleMemoGeneratorTrait>>,
+}
+
+impl Debug for FlexibleMemoGenerator {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("FlexibleMemoGenerator")
+            .field("generator", &self.generator)
+            .finish()
+    }
+}
+
+/// This is the context provided to the flexible memo generator which is set on
+/// the rth memo builder. Each of these fields has a direct correspondence to a
+/// field in the RTHMemoBuilder
+pub struct FlexibleMemoBuilderContext {
```

Comment:
i think this is too complicated, this code will be very hard to follow

instead of implementing a new generator object, can we just add things to the RTHMemoBuilder?